torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -737,13 +496,165 @@ Stmt* LoopNest::lowerToStmt(Tensor* t) {
   return body;
 }
 
-void LoopNest::computeInline(Stmt* s) {
-  // TODO: check if `s` is a body of a loop
-  inlined_functions_.insert(stmt_to_tensor_.at(s)->function());
-}
+class FunctionInliner : public IRMutator {
```

Comment:
One reason we didn't make inliner eager in the first place was the concern that users might choose to inline a function that is transformed later. From talking to several users, if function "A" is inlined into function "B", and some transformations is applied on "A", they expect the later changes show up in the body of "A". That is because from users' logical perspective, "A" and "B" are connected, and changes to "A" should show up in "B".

That being said, I am fine if we choose the other way, if there is a good reason, and we do it consistently. We need to document this clearly why later changes in "A" do not show up in "B".

Reply:
This demonstrates the disconnect that I'm struggling with here. There are no transformations that can be applied to functions in our current design, and before this change, only `computeInline` could be applied to Tensors (this PR changes that though, so no transformations refer to Tensors either). There is no way currently or in any transformation I've seen us discuss for a user to transform "A", they transform loops in the loopnest (reorder, split, merge, fuse, unroll, vectorize), or a single body statement (inline (in this PR), computeAt, rfactor, extract).

What I'm getting at is the concept of a user inlining "A" and then transforming "A" is not expressible in our current design/implementation. I think we need to really be clear about what our abstractions are, who the "user" is in this scenario and what the language they communicate with us is.

With this change I'm bringing `computeInline` in line with our other transformations such as split, rfactor and reorder. In this mode the LoopNest and it's root Statement are equivalent and transformations both refer to and modify the root_stmt's IR directly. This is very different from a model where we transform the input Function or Tensor. I think its fine if we decide thats the route we need to take but I want to stress that's a significant change from where we are now.

Reply:
Let me be a bit more clear with my example: 

C(i, j) = i * j
B(i, j) = C(i, j) + 2
A(i, j) = B(i, j) + 1

This sequence of transformations should work with both "deferred" and "eager" mode:

compute_at(C -> B)
compute_inline(B)

The question is about this:

compute_inline(B)
compute_at(C -> B)

Should the 2nd compute_at be included as part of the the first compute_inline? Either way seems reasonable and could be made self-consistent.

Reply:
I have to confess `computeAt` is the transformation I've looked the least at and I'm not confident about what to do there. 

From a quick read of the code it looks like that second example will fail silently after this change (should probably throw). `computeAt` does not take take Tensors as parameters, it takes a `Stmt*` (in this case would be the store into C) and a `For*`- which in this example is the inner loop of the now empty loop nest that had B and since it can't find usage of C it will bail out. Just checking my understanding is correct here.

I want to repeat that because it's important:  _the concept of `compute_at(C -> B)` is not currently expressible in our API_. We can do instead `compute_at(a store into C, an inner loop that contains a write into B)`. The question of what should happen when somebody wants to inline Tensor A into Tensor B, or `computeAt(C -> B)` is hard to answer because that requires translation into our actual API. If we need to change that to map to what users really want to do, or even to how you think about transformations personally then thats ok, we can have that discussion, but this PR is about making `computeInline` work. Without changing other transformations or abstractions I'm pretty convinced that eager is the only way to do that right now.

I think for the alternative to work we'd need to change computeAt to work on Tensors not statements, change inline back to deferred (not as simple as just not landing this because the old version wont pass the tests), then make all other transformations deferred - because mixing deferred and eager mode is extremely difficult to make work correctly. That's a huge change, a rewrite level change - and I'd be happy to have that discussion again, but it's not I think a good reason to leave `computeInline` in a state where it's silently and unpredictably broken.

Reply:
To answer your question concretely: no, I don't think the second compute at should be part of the first inline, or even succeed at all.